README.md
[![License: The MIT License](https://img.shields.io/badge/license-MIT-green?longCache=true)](https://opensource.org/licenses/MIT)  [![made-with-python](https://img.shields.io/badge/Made%20with-Python-1f425f.svg)](https://www.python.org/)  [![GitHub release](https://img.shields.io/github/release/Naereen/StrapDown.js.svg)](https://GitHub.com/Naereen/StrapDown.js/releases/)

# SystemsRedfishPy

## Cross platform Python tool for provisioning and managing storage systems using the RESTful Redfish/Swordfish API.

#### Copyright (c) 2019 Seagate Technology LLC and/or its Affiliates, All Rights Reserved

## Introduction

***SystemsRedfishPy*** is a command line tool that implements the client side of the Redfish RESTful API for Storage System Management.

Source code files of the SystemsRedfishPy open source project are available to you under [The MIT License](https://opensource.org/licenses/MIT).  The
SystemsRedfishPy project repository is maintained at https://github.com/Seagate.

**Redfish** is the new RESTful API for hardware management defined by the DMTF Scalable Platform Management Forum (SPMF).  It provides a 
modern, secure, multi-node, extendable interface for hardware management. **Swordfish** authored by SNIA provides extensions for handling
storage specific provisioning and management. 

**SystemsRedfishPy** goes beyond common HTTP tools such as curl to provide quick and easy storage management and storage volume creation. This
tool provides simple one-step commands handling multiple Redfish URI requests in order to carry out storage management. The main features of
this package are:
 
* Command line or script file execution
* Set up is a breeze and configuration settings are modified with a **bang** (`!setting [value]`)
* Add new configuration variables via a single line with easy access routines
* Handles multiple command sets for various product brands - `!brand [product]` to switch between command sets 
* Drop in new commands to increase desired functionality - with no modifications to the infrastructure 
* Help text built in to each command file
* Built in unittest for quick regression testing of a Redfish service
* Debug logging and tracing built in for command line usage or unit test cases 
* Class modules to simplify working with URIs and JSON data 

## Background

Seagate Systems (Enterprise Data Solutions (EDS)) provides a Redfish/Swordfish Service that was first released in 2019. This
new service was first featured on the Seagate Indium product line. This package provides a reference client that can be used on
any computer with Python installed. This reference client enables you to perform configuration and maintenance operations on a
storage controller using the Redfish/Swordfish API.

The Redfish API is a standard REST API maintained by [DMTF Redfish](https://www.dmtf.org/standards/redfish). The Swordfish API
is an extension to the Redfish API maintained by [SNIA Swordfish](https://www.snia.org/forums/smi/swordfish).


## Why SystemsRedfishPy?

1. **SystemsRedfishPy** was originally written during the development of the Redfish storage service, helpful for validating
storage service operations and to quickly and easily display JSON data from various URIs.
2. **SystemsRedfishPy** was extended to provide test cases to quickly validate and report on the status of the current
Redfish service version. 
3. **SystemsRedfishPy** provides an example implementation for how a client can execute common storage management functions
like create a RAID disk group, create a storage pool, create a storage volume, and map that volume to a host computer.
4. **SystemsRedfishPy** can also be called from other scripts, used as a command line tool, or execute scripts to quickly provision
storage or check on the health of a storage system. 


## Installation

This project is maintained under [GitHub SystemsRedfishPy](https://github.com/Seagate).

The process to use this client is to clone a copy of the project to your local hard drive.

Clone the project
```
cd base_directory
git clone SystemsRedfishPy.git
cd SystemsRedfishPy
python3 redfishAPI.py
```

## Requirements

Your client computer must have Python3.6 (or later) installed. You will also need network access to the desired controller,
and know the IP Address of the target controller running the Redfish Service. User credentials are required in order to
create a Redfish sessions and provision storage.

Using **SystemsRedfishPy** relies on the packages specified in requirements.txt. But HTML and XML packages are used if you
desire to run unit test cases and produce HTML or XML reports and these are not automatically installed. See the [test document](UNITTEST.md) for more information.

- `pip install -r requirements.txt`

**Note:** You may have to use `pip install --user -r requirements.txt` in some Windows environment.

## Quick Tutorial

This client can run in either an interactive mode, or by parsing a script file. This quick tutorial demonstrates the
interactive mode. All commands entered at the prompt can also be pasted into a text file and run as a script.

Open a terminal window and change directories to the SystemsRedfishPy folder. Run the command 'python redfishAPI.py'
and you will be presented with a '(redfish)' prompt.

As a best practice, it is suggested that you copy redfishAPI.cfg to myconfig.cfg or any name of your choice. Then edit the following properties in that file:
* "dumphttpdata": "True",
* "ipaddress": "[your-service-ip]",
* "password": "[service-password]",
* "username": "[service-username]",
* "http": "http or https",

Additional Properties Note:
* Set `!serviceversion 1` - to work with the earlier version of the Redfish Service that supported ClassesOfService
* Set `!serviceversion 2` - to work with the latest version of the Redfish Service that supports fabrics, composition, etc.
* This can be done at the (redfish) prompt using !serviceversion [1|2]

Then, use python redfishAPI.py -c myconfig.json to run interactive commands using your system configuration settings.

```bash
> python3 redfishAPI.py -c myconfig.json

--------------------------------------------------------------------------------
[2.1.4] Redfish API
--------------------------------------------------------------------------------
[] Run Redfish API commands interactively...

(redfish)
```

There are five main categories of commands that can be entered.
 
### Help

help - provides a list of available commands
help [command name] - provides details on a command
show brands - provides a list of all available brands

### Configuration Commands

There are several configuration commands useful to set up communications and tracing.

| Command            | Description |
| ------------------ | ----------- |
| !dump              | Print out all configuration options. This is useful to learn what settings are available. |
| ![setting] [value] | Change the value for that setting. |

To configure which controller to talk to:

| Command              | Description |
| -------------------- | ----------- |
| !ipaddress 10.235.221.120 | Change all HTTP communications to use this new ip address. |
| !username [name]     | Change the username to `[name]` that is used to log in to the Redfish Service. |
| !password [password] | Change the password to `[password]` that is used to log in to the Redfish Service. |

When running commands, you have several options to help debug issues, and to configure the system. Here is a complete list:

| Command                         | Description |
| ------------------------------- | ----------- |
| !annotate [True,False]          | Provides a banner for every line of script file processed. Default is `True`. |
| !brand [product]                | Specifies the subfolder of commands to use. Default is `systems`, but `example` is provided. |
| !certificatecheck [True,False]  | When False, the URL will be opened using context=ssl._create_unverified_context. Default is `False`. |
| !dump                           | Print out all configuration options. This is useful to learn what settings are available. |
| !dumphttpdata [True,False]      | Display all HTTP data read from the Redfish Service. Useful for additional info. Default is `False`. |
| !dumpjsondata [True,False]      | Display all JSON data read from the Redfish Service. Default is `False`. |
| !dumppostdata [True,False]      | Display all data that is sent via an HTTP POST operation. Default is `False`. |
| !entertoexit [True,False]       | When True, pressing Enter in interactive mode will exit the tool. Default is `False`. |
| !http [https,https]             | Switch between use http:// and https://. Default is `https`. |
| !linktestdelay [seconds]        | How long to delay between URLs when running the 'redfish urls' command. Default is `0`. |
| !ipaddress [ip]                 | Change all HTTP communications to use this new ip address. |
| !port [port]                    | Change all HTTP communications to use this new port value, default is 80. |
| !password [password]            | Change the password to `[password]` that is used to log in to the Redfish Service. |
| !serviceversion [1,2]           | Specify the Redfish Service version. This changes command behavior based on supported schemas. Default is `2`. |
| !showelapsed [True,False]       | Display how long each command took. Default is `False`. |
| !trace [4-7]                    | Turn on additional tracing. 4=DEFAULT, 5=VERBOSE, 6=DEBUG, 7=TRACE. Default is `4`. |
| !urltimeout [seconds]           | How long to wait for a URL request before timing out. Default is `30`. |
| !usefinalslash [True,False]     | When True, all Redfish URIs will have a slash as the final character in the URL. Default is `True`. |
| !username [name]                | Change the username to `[name]` that is used to log in to the Redfish Service. |


### Redfish Commands

Most commands require that you establish a session with the target Redfish Service. To do so, use 'create session'.
This command will use the configuration settings, listed above, such as ipaddress, username, and password and attempt to
establish a session.

```bash
--------------------------------------------------------------------------------
[2.1.4] Redfish API
--------------------------------------------------------------------------------
[] Run Redfish API commands interactively...

(redfish) create session
   -- ServiceVersion: 2
   -- IP Address    : https://<ip>
   -- Discovered: Root                      >> /redfish/v1
   -- Discovered: Chassis                   >> /redfish/v1/Chassis/
   -- Discovered: CompositionService        >> /redfish/v1/CompositionService/
   -- Discovered: Fabrics                   >> /redfish/v1/Fabrics/
   -- Discovered: Managers                  >> /redfish/v1/Managers/
   -- Discovered: SessionService            >> /redfish/v1/SessionService/
   -- Discovered: Systems                   >> /redfish/v1/Systems/
   -- Discovered: Tasks                     >> /redfish/v1/TaskService/
   -- Discovered: UpdateService             >> /redfish/v1/UpdateService/
   -- Discovered: Sessions                  >> /redfish/v1/SessionService/Sessions/
   -- Discovered: metadata                  >> /redfish/v1/$metadata/
   -- Discovered: odata                     >> /redfish/v1/odata/

[] Redfish session established (4:<sessionkey>)

[] Elapsed time: 0m 2s to execute command
```

The main redfish commands are used for debugging or learning more about the data returned by the Redfish Service.

For example:

| Command                     | Description |
| --------------------------- | ----------- |
| redfish json [url]          | Display the JSON data returned from a GET to [url]. Errors are also reported. |
| redfish urls [optional url] | Traverse every URL reported by the service, validate them, and produce a report. If no optional url is specified, then traversing starts with '/redfish/v1'. |


### Systems Commands

The system commands use the Redfish Service API to create volumes, display disks, and other configuration and
maintenance items. Use 'help [command]' to display additional examples. 

For example:

```bash
(redfish) show disks - will display all disk drives in the system
(redfish) show pools - will display all configured virtual pools
(redfish) show volumes - will display all configured volumes
(redfish) create diskgroup name=dgA01 disks=0.7,0.8 pool=A level=raid1 - to create a new RAID1 disk group
(redfish) create volume name=TestVol01 size=100000000000 pool=A - to create a new volume
```

### HTTP Commands

The HTTP commands allow you to do native HTTP GET, DELETE, POST, and PATCH commands.

As a note, most Redfish commands require that you establish a session with the target Redfish Service. If you use
the 'http post' command to create a session, you should also call 'save session [id] [key]' so that all following
http commands can use the session credentials.
 
| Command                                       | Description |
| --------------------------------------------- | ----------- |
| http get [url]                                | Perform an HTTP GET operation on a URI |
| http post [url] [json or filename]            | Perform an HTTP POST operation on a URI, sending the specified JSON data |
| http patch [url] [json or filename]           | Perform an HTTP PATCH operation on a URI, sending the specified JSON data |
| http delete [url]                             | Perform an HTTP DELETE operation on a URI |
| http push [url] [imagefile] [<]json or file]  | Perform an HTTP POST operation to transfer a file to a URI |

Examples:

```bash
(redfish) http get /redfish
(redfish) http get /redfish/v1
(redfish) http post /redfish/v1/SessionService/Sessions { "UserName": "[username]", "Password": "[password]" }
(redfish) save session [id] [key]
(redfish) http get /redfish/v1/Chassis
(redfish) http get /redfish/v1/Managers
(redfish) http post /redfish/v1/SessionService/Sessions credentials.json
```
### Get Log Commands

The get logs commands allow you retrieve controller or drive logs. Here are examples of executing each.

```bash
(redfish) get logs component=controller

++ get logs: /redfish/v1/Systems/00C0FF437ED5/LogServices/controller_a/Actions/LogService.CollectDiagnosticData...
++ POST get logs (controller, CollectControllerLog)

   >> Monitor task 'getlogs-task' for Completed
      == TaskState: 'Running' (sleep 30)
      == TaskState: 'Running' (sleep 30)
      == TaskState: 'Running' (sleep 30)
      == TaskState: 'Running' (sleep 30)
      == TaskState: 'Running' (sleep 30)
      == TaskState: 'Running' (sleep 3)
      == TaskState: 'Running' (sleep 3)
      == TaskState: 'Running' (sleep 3)
      == TaskState: 'Running' (sleep 3)
      == TaskState: 'Running' (sleep 3)
      == TaskState: 'Completed'

++ POST get logs (controller, DownloadLogData)
   -- Status        : 200
   -- Reason        : OK
   -- Download complete to 'logfile.zip'
```

An example of retrieving disk drive logs.

```bash
(redfish) get logs component=drive

```

## Redfish Tutorials

| Tutorial                       | Description |
| ------------------------------ | ----------- |
| [tutorial-redfish-service-v1.md](tutorial-redfish-service-v1.md) | Explore Redfish API operations using the Seagate Systems Redfish Service v1. |


## Design

If you want to make changes to this reference client, there is a [design document](DESIGN.md) that provides an overview
of how to make changes and add new commands. The main system design allows you to add commands, and help for
commands, without having to change any of the underlying core files. The only step needed is to add your new
command to the 'commands' folder using the prescribed template.


## Unit Testing

If you want to run unit testing, the [unit test document](UNITTEST.md) provides an overview of how to install
the reporting packages, run unit tests, and also add new unit tests.


CHANGELOG.md
# Change Log

## [2.5.2] - 2023-06-06
- Handle ServiceRoot services that do not have a `@odata.id` included

## [2.5.1] - 2022-12-07
- Updated the types of drive logs available in 'get logs'.

## [2.5.0] - 2022-11-15
- Added support of Redfish event listener in the core
- Added command to start/ stop Redfish event listener

## [2.4.2] - 2022-10-06
- Log failures to load invalid JSON configuration files

## [2.4.1] - 2022-09-08
- Install Atheris from Google
- Fuzz the command input using Atheris from Google
- Fix crash caused by surrogate pair characters

## [2.4.0] - 2022-08-24
- Switched to using python requests library for all http operations, urllib has SSL limitations
- UrlAccess.process_request() does not work with formatted JSON data using json.dumps()
- Adjustments were required for working with the request response.headers
- Adjust the `help` command to show help for brand and common commands
- Display call stack for command exceptions when trace level is DEBUG or higher

## [2.3.0] - 2022-08-01
- Created new 'common' command folder. Brand is searched first, then common.
- Added new 'show brands' command which displays all brands and a command count.
- Moved many commands from 'systems' and 'example' to 'common'
- Minor bug correction in 'create session'
- Created a new Open BMC brand (obmc) as a starting point
- Added new core/display.py with max_width() function

## [2.2.6] - 2022-08-16
- Skipped warning message for zip content type when using "get logs component=controller".
- Removed use of the undefined "context" variable, this ensures "http push <firmware_file>" does not exit with an error.

## [2.2.5] - 2022-06-06
- Replaced `mcip` configuration setting with `ipaddress` to allow clearer integration with other Redfish services.
- Add `port` as a new configuration setting. By default, this is set to 80 but any valid Redfish service port will work.
- Changed config option 'httpbasicauth' to 'basicauth' to reduce typing and make the setting more visible.

## [2.2.4] - 2022-04-06
- New `load config` command making it easier to switch from one config to another.
- Correction to the StorageGroup steps [Tutorial for Redfish Service v2.4.20](tutorial-redfish-service-v2.md).
- Changed Trace.NONE to Trace.ALWAYS which will always print the trace entry.

## [2.2.3] - 2022-02-03
- Small correction to [Tutorial for Redfish Service v2.4.20](tutorial-redfish-service-v2.md).

## [2.2.2] - 2022-01-31
- New [Tutorial for Redfish Service v2.4.20](tutorial-redfish-service-v2.md).
- Several changes to json/* example files to update them to the Redfish Service v2.x.x format.
- Correction to handling JSON data on the command line.

## [2.2.1] - 2021-11-15
- Updates to JSON files used to patch StorageGroups since parameter AccessCapabilities is now AccessCapability

## [2.2.0] - 2021-10-28
- New `version` command to display the version of this tool
- Renamed example configuration `redfishAPI.json` file to `redfishAPI.cfg`
- Modified `.gitignore` to ignore all `*.cfg` files (suggest renaming `.json` config files to `.cfg`)
- Added `show accounts` to use /redfish/v1/AccountService to display all Accounts
- Added `create account` to use /redfish/v1/AccountService/Accounts to create a new user account
- Added `update account` to use /redfish/v1/AccountService/Accounts to update a user account
- Added `delete accounts` to use /redfish/v1/AccountService/Accounts to remove one or more user accounts
- Added `create snapshot` to use {VolumeId}/Actions/Volume.CreateReplicaTarget to create a volume snapshot
- New `assert [operator]` command to aide with http url validation testing (see `scripts\test_invalid_uris.rfs`)
- Added new label `$httpstatus` which is used by the `assert` command

## [2.1.9] - 2021-9-7
- Expanded `redfish version` to display Redfish Specification and Redfish Service versions.
- New common urlAccess.print_status() to print HTTP status code, reason, and context
- New json files for StorageGroup testing (that require modifications for each host)

## [2.1.8] - 2021-8-24
- Corrections to HTTP Basic Auth handling
- Corrections to markdown links and configuration text

## [2.1.7] - 2021-7-8
- Changed default urltimeout to 300 to properly handle log file retrieval.
- Allow the user to specify a filename when running `get logs`
- Changed the help for create diskgroup to list 'raid10'

## [2.1.6] - 2021-6-7
- Added a new print to display the current trace level during start up.
- Corrected the urlAccess handling of url data that is either bytes or str.

## [2.1.5] - 2021-5-28
- Corrected an issue with getting configuration settings, which affected commands like `redfish urls`
- Added the capability of handling a `hostname` as well as an IP Address for the storage array access.

## [2.1.4] - 2021-4-20
- Added `get logs` to retrieve controller and disk drive logs
- Added `cs` alias as a short cut to run `create session`
- Added `decode` flag for UrlAccess.process_request() to handle retrieval of log files
- The URI discovery tag `ActiveController` is now `StorageActiveController` to relate better to the URI
- Added URI discovery service for `SystemsLogServices`
- Added the display of IP Address when creating a session for easier debug

## [2.1.3] - 2021-3-19
- Added `help configuration` to show how to use dump and set config values, also sorted help command output

## [2.1.2] - 2021-3-18
- Added `show tasks`

## [2.1.1] - 2021-2-3
- Added [Troubleshooting Guide](troubleshooting.md)
- Corrections and enhancements to `show thermal`, `create volume`, `compose volume`, `create diskgroup`,`create storagegroup`,

## [2.1.0] - 2021-1-22
- Small correction to how the Disks URL is discovered under serviceversion=1. This version will be pushed to github.

## [2.0.7] - 2021-1-22
- Updated the `http push` command to work correctly for Redfish Service firmware uploads.

## [2.0.6] - 2021-1-14
- New example JSON files for volume composition under json/

get_logs.py
#
# Do NOT modify or remove this copyright and license
#
# Copyright (c) 2019 Seagate Technology LLC and/or its Affiliates, All Rights Reserved
#
# This software is subject to the terms of the MIT License. If a copy of the license was
# not distributed with this file, you can obtain one at https://opensource.org/licenses/MIT.
#
# ******************************************************************************************
#
# get_logs.py 
#
# ******************************************************************************************
#
# @command get logs
#
# @synopsis Retrieve controller or disk drive logs and store in logs.zip.
#
# @description-start
#
# 'get logs component=[controller|drive] logtype=[disk|diskfarm|etc] drivenumber=[enclosure.slot]
#
# Note:
#     This command is only valid for Redfish Service Version 2.
#
# Parameters:
#     'component' - Required
#         options: controller, drive
#
#     'logtype' - Required for 'drive', not used for 'controller'
#         options: disk, diskfarm, diskibt, disksm2, diskidd, diskfarmlct, diskfarmfield,
#                 diskfarmts1, diskfarmts2, diskfarmsf1, diskfarmsf2, diskfarmsf3,
#                 diskfarmsf4, diskfarmfactory
#
#     'drivenumber' - Required for 'drive', not used for 'controller'
#         options: enclosure.slot, for example 0.24
#
#     Note: Not all drive log types are available for all drives. If a drive log type is
#           supplied that is not supported by the drive, the command will fail.
#
# Examples:
#     get logs component=controller
#     get logs component=drive logtype=diskfarm drivenumber=0.1
#
# @description-end
#

import json
from commands.commandHandlerBase import CommandHandlerBase
from core.jsonBuilder import JsonBuilder, JsonType
from core.redfishSystem import RedfishSystem
from core.trace import TraceLevel, Trace
from core.urlAccess import UrlAccess, UrlStatus
import time
import zipfile

# Display the results of a get logs call
def display_log_results(link, log_filename, display_contents):
    error_filename = 'ErrorMessage.txt'
    error_exists = False

    Trace.log(TraceLevel.INFO, "   -- {0: <14}: {1}".format('Status', link.urlStatus))
    Trace.log(TraceLevel.INFO, "   -- {0: <14}: {1}".format('Reason', link.urlReason))

    if link.urlStatus == 200 and link.urlData != None:
        outF = open(log_filename, 'wb')
        outF.write(link.urlData)
        outF.close()
        Trace.log(TraceLevel.INFO, "   -- Download complete to '{}'".format(log_filename))

        if display_contents:
            Trace.log(TraceLevel.INFO, "")
            Trace.log(TraceLevel.INFO, "   -- Contents of zip file '{}':".format(log_filename))

        # Display contents of zip file and check for error file
        zip_file = zipfile.ZipFile(log_filename, 'r')
        for name in zip_file.namelist():
            if name == error_filename:
                error_exists = True
            if display_contents:
                Trace.log(TraceLevel.INFO, "      ** {}".format(name))

        # Extract and display error file if it exists
        if error_exists:
            zip_file.extractall()
            Trace.log(TraceLevel.INFO, "")
            Trace.log(TraceLevel.INFO, "   -- Contents of file '{}':".format(error_filename))
            with open(error_filename) as f:
                lines = f.read().splitlines()
                for line in lines:
                    if len(line.strip()) > 0:
                        Trace.log(TraceLevel.INFO, "      || {}".format(line))
        zip_file.close()
    else:
        Trace.log(TraceLevel.INFO, "ERROR: get logs POST returned no URL data")


# The get logs call returns a task id which needs to be monitored until
# until the TaskState is 'Completed', once completed, execute another POST
# to retrieve the log file.
def process_task_status(redfishConfig, link, log_filename):

    if link.urlStatus != 201:
        # First POST failed, don't proceed further
        Trace.log(TraceLevel.INFO, "   -- {0: <14}: {1}".format('Status', link.urlStatus))
        Trace.log(TraceLevel.INFO, "   -- {0: <14}: {1}".format('Reason', link.urlReason))
        return

    taskid = ''
    if (link.valid and link.jsonData is not None and 'Id' in link.jsonData):
        taskid = link.jsonData['Id']

    Trace.log(TraceLevel.INFO, "")
    Trace.log(TraceLevel.INFO, "   >> Monitor task '{}' for Completed".format(taskid))

    # Poll the task status until it is Completed
    completed = 0
    delay = 30
    counter = 1
    uri = RedfishSystem.get_uri(redfishConfig, 'Tasks') + 'Tasks/' + taskid
    while completed == 0:
        link2 = UrlAccess.process_request(redfishConfig, UrlStatus(uri), 'GET', True, None)
        if link2.urlStatus == 200:
            if (link2.valid and link2.jsonData is not None and 'TaskState' in link2.jsonData):
                if link2.jsonData['TaskState'] == 'Completed':
                    Trace.log(TraceLevel.INFO, "      == TaskState: '{}'".format(link2.jsonData['TaskState']))
                    completed = 1
                else:
                    Trace.log(TraceLevel.INFO, "      == TaskState: '{}' (sleep {})".format(link2.jsonData['TaskState'], delay))
                    time.sleep(delay)
                    # Reduce the sleep delay after 2m 30s
                    if counter == 5:
                        delay = int(delay / 10)
                    counter += 1
            else:
                Trace.log(TraceLevel.INFO, "      == ERROR: Failed to obtain task status")
                completed = 2
        else:
            Trace.log(TraceLevel.INFO, "      == {0: <14}: {1}".format('Status', link2.urlStatus))
            Trace.log(TraceLevel.INFO, "      == {0: <14}: {1}".format('Reason', link2.urlReason))
            completed = 3

    # Retrieve and store the log file
    if completed == 1:
        JsonBuilder.startNew()
        JsonBuilder.newElement('main', JsonType.DICT)
        JsonBuilder.addElement('main', JsonType.STRING, 'DiagnosticDataType', 'OEM')
        JsonBuilder.addElement('main', JsonType.STRING, 'OEMDiagnosticDataType', 'DownloadLogData')

        Trace.log(TraceLevel.INFO, "")
        Trace.log(TraceLevel.INFO, "++ POST get logs (controller, DownloadLogData)")
        link3 = UrlAccess.process_request(redfishConfig, UrlStatus(link.url), 'POST', True, json.dumps(JsonBuilder.getElement('main'), indent=4), decode=False)
        display_log_results(link3, log_filename, display_contents=False)


################################################################################
# POST Request Body Examples
################################################################################
#
# Example JSON POST request data for CollectControllerLog
# {
#     "DiagnosticDataType": "OEM",
#     "OEMDiagnosticDataType": "CollectControllerLog"
# }
#
# Example JSON POST request data for GetDriveLog
# {
#     "DiagnosticDataType": "OEM",
#     "OEMDiagnosticDataType": "GetDriveLog"
#     "DriveLogType": "diskfarm"
#     "DriveNumber": "0.1"
# }
#

################################################################################
# CommandHandler
################################################################################
class CommandHandler(CommandHandlerBase):
    """Command - get logs"""
    name = "get logs"
    command = ""

    @classmethod
    def prepare_url(self, redfishConfig, command):
        self.command = command
        # Example: /redfish/v1/Systems/00C0FF472054/LogServices/controller_a/Actions/LogService.CollectDiagnosticData
        uri = RedfishSystem.get_uri(redfishConfig, 'SystemsLogServices') + 'Actions/LogService.CollectDiagnosticData'
        return (uri)

    @classmethod
    def process_json(self, redfishConfig, url):
        Trace.log(TraceLevel.INFO, "")
        Trace.log(TraceLevel.INFO, "++ get logs: {}...".format(url))

        if (redfishConfig.get_version() < 2):
            Trace.log(TraceLevel.INFO, "ERROR: get logs is only supported in service version 2: (ServiceVersion={})".format(redfishConfig.get_version()))
            return

        jsonType, component = JsonBuilder.getValue('component', self.command)
        if (jsonType is JsonType.NONE):
            Trace.log(TraceLevel.INFO, "ERROR: get logs requires that you provide a 'component' (controller, drive)")
            return

        # From the command, build up the required JSON data

        logtype = ''
        drivenumber = ''

        JsonBuilder.startNew()
        JsonBuilder.newElement('main', JsonType.DICT)

        if component == 'controller':
            JsonBuilder.addElement('main', JsonType.STRING, 'DiagnosticDataType', 'OEM')
            JsonBuilder.addElement('main', JsonType.STRING, 'OEMDiagnosticDataType', 'CollectControllerLog')

        elif component == 'drive':
            JsonBuilder.addElement('main', JsonType.STRING, 'DiagnosticDataType', 'OEM')
            JsonBuilder.addElement('main', JsonType.STRING, 'OEMDiagnosticDataType', 'GetDriveLog')
            # DriveLogType
            jsonType, logtype = JsonBuilder.getValue('logtype', self.command)
            if (jsonType is not JsonType.NONE):
                JsonBuilder.addElement('main', JsonType.STRING, 'DriveLogType', logtype)
            # DriveNumber
            jsonType, drivenumber = JsonBuilder.getValue('drivenumber', self.command)
            if (jsonType is not JsonType.NONE):
                JsonBuilder.addElement('main', JsonType.STRING, 'DriveNumber', drivenumber)

        else:
            Trace.log(TraceLevel.INFO, "ERROR: get logs unsupported 'component' type ({})".format(component))
            return

        # Execute the required HTTP operations to perform the log pull

        log_filename = 'logfile.zip'

        if component == 'controller':
            Trace.log(TraceLevel.INFO, "++ POST get logs (controller, CollectControllerLog)")
            link = UrlAccess.process_request(redfishConfig, UrlStatus(url), 'POST', True, json.dumps(JsonBuilder.getElement('main'), indent=4), decode=False)
            process_task_status(redfishConfig, link, log_filename)

        else:
            Trace.log(TraceLevel.INFO, "++ POST get logs (drive, {}, {})".format(logtype, drivenumber))
            link = UrlAccess.process_request(redfishConfig, UrlStatus(url), 'POST', True, json.dumps(JsonBuilder.getElement('main'), indent=4), decode=False)
            display_log_results(link, log_filename, display_contents=True)

    @classmethod
    def display_results(self, redfishConfig):
        # Nothing to do in this case
        Trace.log(TraceLevel.INFO, " ")
